doc/book/reference/migration/to-v1-1.md

```diff
@@ -0,0 +1,688 @@
+    ],
+    'middleware_pipeline' => [
+        'always' => [
+            'middleware' => [
+                OriginalMessages::class, // <----- Add this entry
+                Helper\ServerUrlMiddleware::class,
+                /* ... */
+            ],
+            'priority' => 10000,
+        ],
+
+        /* ... */
+    ],
+];
+```
+
+If you are [programmatically creating your pipeline](https://mwop.net/blog/2016-05-16-programmatic-expressive.html),
+use the following:
+
+```php
+$app->pipe(OriginalMessages::class);
+/* all other middleware */
+```
+
+### Identifying and fixing getOriginal calls
+
+To help you identify and update calls in your own code to the `getOriginal*()`
+methods, we provide a tool via the [zendframework/zend-expressive-tooling](https://github.com/zendframework/zend-expressive-tooling)
+package, `vendor/bin/expressive-migrate-original-messages`.
+
+First, install the tooling package; since the tooling it provides is only
+useful during development, install it as a development requirement:
+
+```bash
+$ composer require --dev zendframework/zend-expressive-tooling
+```
+
+Once installed,  you can execute the tool using:
+
+```bash
+$ ./vendor/bin/expressive-migrate-original-messages
+```
+
+Passing the arguments `help`, `--help`, or `-h` will provide usage information;
+in most cases, it will assume sane defaults in order to run its scans.
+
+The tool updates calls to `getOriginalRequest()` and `getOriginalUri()` to
+instead use the new request attributes that the `OriginalMessages` middleware
+injects:
+
+- `getOriginalRequest()` becomes `getAttribute('originalRequest', $request)`
+- `getOriginalUri()` becomes `getAttribute('originalUri', $request->getUri())`
+
+In both cases, `$request` will be replaced with whatever variable name you used
+for the request instance.
+
+For `getOriginalResponse()` calls, which happen on the response instance, the
+tool will instead tell you what files had such calls, and detail how you can
+update those calls to use the `originalResponse` request attribute.
+
+## Programmatic middleware pipelines
+
+With Expressive 1.0, we recommended creating middleware pipelines and routing
+via configuration. Starting with 1.1, we recommend *programmatic creation of
+pipelines and routing*.
+
+Programmatic pipelines exercise the existing Expressive API. Methods include:
+
+- `pipe()` allows you to pipe middleware for the pipeline; this can optionally
+  take a `$path` argument. (If one argument is present, it is assumed to be
+  middleware; with two arguments, the first argument is the `$path`.) Paths are
+  literal URI path segments. If the incoming request matches that segment, the
+  middleware will execute; otherwise, it will not. These can be used to provide
+  sub-applications with their own routing.
+
+- `pipeRoutingMiddleware()` is used to pipe the internal routing middleware into
+  the pipeline.
+
+- `pipeDispatchMiddleware()` is used to pipe the internal dispatch middleware into
```

Comment:
In 1.1 (based off of current master) it still is; 2.0 (based off of current develop) makes it a separate class. ;-)